### ISSUE TITLE
A exception "The Content-MD5 you specified did not match what we received", what can I do?

### ISSUE THREAD

**Issue by liujiayi771** [aws/aws-sdk-java#665]
Today, I get a new exception I never get before when I test my application that upload file from ec2 to s3. The content is:
```
Exception in thread "Thread-1" com.amazonaws.services.s3.model.AmazonS3Exception: The Content-MD5 you specified did not match what we received. (Service: Amazon S3; Status Code: 400; Error Code: BadDigest; Request ID: 972CB8E04388AB20), S3 Extended Request ID: T7bmFnQ2RlGWlJD+aGYfTy97XZw88pbQrwNB8YCezSjyq6O2joxHRP/6ko+Q2zZeGewkw4x/90k=
	at com.amazonaws.http.AmazonHttpClient.handleErrorResponse(AmazonHttpClient.java:1383)
	at com.amazonaws.http.AmazonHttpClient.executeOneRequest(AmazonHttpClient.java:902)
	at com.amazonaws.http.AmazonHttpClient.executeHelper(AmazonHttpClient.java:607)
	at com.amazonaws.http.AmazonHttpClient.doExecute(AmazonHttpClient.java:376)
	at com.amazonaws.http.AmazonHttpClient.executeWithTimer(AmazonHttpClient.java:338)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:287)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:3676)
	at com.amazonaws.services.s3.AmazonS3Client.putObject(AmazonS3Client.java:1439)
	at com.amazonaws.services.s3.transfer.internal.UploadCallable.uploadInOneChunk(UploadCallable.java:131)
	at com.amazonaws.services.s3.transfer.internal.UploadCallable.call(UploadCallable.java:123)
	at com.amazonaws.services.s3.transfer.internal.UploadMonitor.call(UploadMonitor.java:139)
	at com.amazonaws.services.s3.transfer.internal.UploadMonitor.call(UploadMonitor.java:47)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

```
What can I do to fix this bug? I use the same code as before in my application.

**Comment by varunnvs92** [comment]
Hey, Can you provide us with sample code you are using?

S3 calculates the checksum at the beginning of your upload. Make sure you are not changing it during your upload.

**Comment by liujiayi771** [comment]
Hearing what you said, I think I have found the reason. I use muti-threading progran—one thread generating the file I want to upload, another uploding the file to S3. Sometimes, the file has not been generated completely and been uploaded at the same time, so that the content of the file is actually different from the beginning of the uploading which caused the changing of md5. I think I should guarantee the mutually exclusive resource usahe. Thank you very much!

**Comment by varunnvs92** [comment]
Cool! Feel free to contact us if you have more questions.

**Comment by ilirlazaj07** [comment]
So how do you do it pratically...in JAVA ?

**Comment by varunnvs92** [comment]
@ilirlazaj07 
Didn't grok the question fully. Can you give me more details? Thanks.

**Comment by ilirlazaj07** [comment]
Im trying to upload a file in a local customized instance of S3. I'm using Java SDK but when I try to PUT the file the console says: **Exception in thread "Thread-1" com.amazonaws.services.s3.model.AmazonS3Exception: The Content-MD5 you specified did not match what we received. (Service: Amazon S3; Status Code: 400; Error Code: BadDigest; Request ID: null), S3 Extended Request ID: null**

When I try doing the same thing via the AWS CLI everything works fine

**Comment by ilirlazaj07** [comment]
My Code: public class TEST {

    public static void main(String[] args) {

        BasicAWSCredentials awsCreds = new BasicAWSCredentials("default", "default");

        AmazonS3 s3Client = AmazonS3ClientBuilder.standard()
                .withCredentials(new AWSStaticCredentialsProvider(awsCreds)).
                withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration("http://127.0.0.1:8089", "eu-west-1")).
                build();
        
        System.out.println("Uploading a new object to S3 from a file\n");
        File file = new File("C:\\test.txt");
        s3Client.putObject(new PutObjectRequest(
                "s3://fake", "", file));

    }

}

**Comment by aj07mm** [comment]
yeah, not sure about the answer. Can you paste here the code you did to solve that?

**Comment by jazzyjackson** [comment]
Hi! I just ran into this error message, but I'm not doing anything asynchronous where the file content would change during the upload.

Instead, I'm creating an empty buffer in memory using StringIO.StringIO(): just because pandas can write to a buffer, and s3.putObject can upload a buffer.

BUT the thing about buffers is the 'byte offset' changes as you read or write. Long story short, I was able to clear this error by using .seek(0) to reset my stream to the beginning before starting the s3 upload. 

```py
samples = pandas.DataFrame(numpy.random.rand(valid.rows, valid.cols))
fileBody = StringIO.StringIO()
samples.to_csv(
    fileBody, # note that pyvalidate currently writes bytes! not unicode... python2 thing :(
    index=valid.get('labels', False),
    header=valid.get('labels', False),
)
fileBody.seek(0)
valid.s3Object.put(Body=fileBody)
```

Hope that helps.
